### ISSUE TITLE
New checks for template analyzer

### ISSUE THREAD

**Issue by vlada-shubina** [sayedihashimi/dotnet-new-web#42]
We would like to include the following checks to template analyzer tools: 

- warning: if the user puts the email in the author field (discussion ongoing) 
- warning: if the user starts primary output with "/" (https://github.com/dotnet/templating/issues/2682)
- warning: if applying forms on `replace` and `filerename` values doesn't provide different result (https://github.com/dotnet/templating/issues/2621)

**Comment by sayedihashimi** [comment]
Added support for showing an error when `primaryOutputs.path` starts with a `/` or `\` in 9044c5f063f1f86c127dd187a7f9eacd485193d8.

I'm not sure about the first bullet point. I don't see that as being warning, or error, level. If you can point me to the discussion the additional context may be helpful.

Can you give more info on the third bullet.

**Comment by vlada-shubina** [comment]
@sayedihashimi thank you for addressing it.  
The first point is not required anymore. We discussed it with legal and they confirmed that it's OK to use email in the author field. 

The third point applies to advanced template configuration. In general, when the template author is using forms on symbols applying the forms to the values of `replace` or `file rename` should produce the different result, otherwise it is not possible to determine the form to use and random selection will be made.

Example
```
  "symbols": {
	"test": {
      "type": "parameter",
      "datatype": "string",
      "description": "test",
      "defaultValue": "MyValue",
	  "replaces": "testreplace",
	  "fileRename": "testreplace",
	  "forms": {
        "global": [ "identity", "lc"]
      }	  
    },
    "forms": {
    "lc": {
      "identifier": "lowerCase"
    }
}
```
For `testreplace` both `identity` and `lc` form is `testreplace`. When user specifies `MyTestProject` value for `test` symbol, it's not certain whether `testreplace` will be replaced with `MyTestProject` or `mytestproject`. 
Therefore the author should choose better `replace` value which differs for all forms provided: for example `TestReplace` (identity: TestReplace, lc: testreplace). Then for example above:
- `TestReplace` will be replaced with `MyTestProject`
- `testreplace` will be replaced with `mytestproject`

Hope it helps.

**Comment by sayedihashimi** [comment]
@vlada-shubina sorry for the delay here. I see how this could be useful to add as an analyzer. What I'm not sure of is, how can the analyzer know the different forms that will be produced by the forms declared. Is there an easy way for me to generate the forms from the analyzer?

**Comment by vlada-shubina** [comment]
The easiest way imo is to use `Microsoft.TemplateEngine.Orchestrator.RunnableProjects` (if you are not using it already), load `SimpleConfigModel` through one of the constructors, go through all `ParameterSymbol`s and apply `Forms` to `Replace` and `FileRename` values. Forms itself are defined in `SimpleConfigModel.Forms` with `Process` method allowing to apply the form to string.

Note: this implementation is now internal, however I can add your assembly in friendly assemblies list, or define public APIs for those. This can also be used to identify other errors, for example if form defined for `ParameterSymbol` is not available at all, or `Forms` defined in `SimpleConfigModel` are not correct.

**Comment by sayedihashimi** [comment]
Thanks @vlada-shubina. Currently the analyzer it totally decoupled from the TemplateEngine. I wonder if it would be better to do a re-write and integrate TemplateEngine bits so that we can have a better analysis. What do you think?

**Comment by vlada-shubina** [comment]
yes, I think it would more efficient to use `SimpleConfigModel` as then you have the object model, and not json, also with access to some further functionality as `Forms`. Also this allows to ensure that analyzer and template engine are interpreting json in the same way.

I created https://github.com/dotnet/templating/issues/4142 to review the APIs and make them public. Meanwhile I can make internals visible to the assembly to allow to proceed.

The only thing that is worth to be checked additionally is invalid json properties, as `SimpleConfigModel` would skip those.
